Turn by turn processing

br648 opened the pull request:
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [x] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

This PR aims to provide turn by turn instructions to a traveler. It will inform the traveler of upcoming and immediate instructions based on the information provided in a step.

A list of points for the leg are created from the leg's geo points, start point, end point and step points. Starting with the geo points, the step points are injected based on being between two points. Any step points not added, will be injected based on proximity to a geo point. The final list of points (If b = beginning, p = point on leg, S = step and e = end) would be something like:

b-p-S-p-p-p-p-p-p-S-p-p-S-p-p-p-p-p-S-e

The travellers poisition is first snapped to one of these points to define a position on the leg. Using this as a starting point, the following points are traversed until a "step point" is found. The instruction then provided will be dependent on how far the traveller is from this step. Less than two meters an immediate instruction is given, less than ten meters an upcoming instruction is given, else no instruction is given.

br648 committed:
refactor(Initial commit):

br648 committed:
refactor(Further updates to produce turn-by-turn instructions):

br648 committed:
refactor(README.md): Updated to include latest params.

br648 committed:
refactor(ManageTripTracking.java): Removed redundant comment/code

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/triptracker/ManageLegTraversal.java:
```diff
@@ -29,38 +29,38 @@ private ManageLegTraversal() {
      * is a straight line.
      */
-    public static Segment getSegmentFromPosition(Leg leg, Coordinates currentCoordinates) {
+    public static TripSegment getSegmentFromPosition(Leg leg, Coordinates currentCoordinates) {
```
Maybe rename to `LegSegment`?

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/triptracker/TripInstruction.java:
```diff
@@ -1,30 +1,217 @@
+    }
+
+    /**
+     * The trip instruction can only be provided if the trip is on or close to schedule.
```
I think you meant to write "if the traveler is close to the indicated route" (i.e. not deviated).

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/utils/Coordinates.java:
```diff
@@ -43,4 +55,9 @@ public boolean equals(Object o) {
+
+    @Override
+    public String toString() {
+        return lat + "," + lon + ",";
```
What is the trailing comma for?

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/triptracker/TripInstruction.java:
```diff
@@ -1,30 +1,217 @@
+    }
+
+    /**
+     * Get instruction based distance and related step instruction.
```
Can you add an example of how that should look, please. I would assume this is a formatted code as opposed to human-readable content?

binh-dam-ibigroup reviewed (commented):
Haven't got into the weeds yet...

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/triptracker/TripStatus.java:
```diff
@@ -65,23 +64,22 @@ public enum TripStatus {
-                    .plusSeconds((long) getSegmentTimeInterval(segmentFromPosition));
-                return currentTime.isBefore(segmentCenterTime) ? TripStatus.BEHIND_SCHEDULE : TripStatus.AHEAD_OF_SCHEDULE;
+                    .plusSeconds((long) getSegmentTimeInterval(travelerPosition.tripSegmentFromPosition));
+                return travelerPosition.currentTime.isBefore(segmentCenterTime)
```
Something is odd here, when following an itinerary very slowly starting before the start time, this will report me from being "Behind schedule" to being "ahead of schedule". This should probably be the other way around.

binh-dam-ibigroup commented on src/test/java/org/opentripplanner/middleware/controllers/api/ManageLegTraversalTest.java:
```diff
@@ -118,20 +130,104 @@ private static Stream<Arguments> createTrace() {
+        return String.format("%s%s", prefix, step.relativeDirection);
+    }
+
+    private static TripSegment createTripSegment(Coordinates point, double distance, double bearing, boolean oppositeDirection) {
```
What unit is `distance` in?

binh-dam-ibigroup commented on src/test/java/org/opentripplanner/middleware/controllers/api/ManageLegTraversalTest.java:
```diff
@@ -51,16 +61,18 @@ void canTrackTrip(String time, double lat, double lon, TripStatus expected) {
-        TripStatus tripStatus = ManageTripTracking.getTripStatus(trackedJourney, busStopToJusticeCenterItinerary);
+        TravelerPosition travelerPosition = new TravelerPosition(trackedJourney, busStopToJusticeCenterItinerary);
+        TripStatus tripStatus = TripStatus.getTripStatus(travelerPosition);
+        String instruction = TripInstruction.getInstruction(tripStatus, travelerPosition);
```
This variable appears unused.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/triptracker/TripInstruction.java:
```diff
@@ -1,30 +1,217 @@
+                    Step nextStep = expectedLeg.steps.get(nearestStep.stepIndex + 1);
+                    return getInstruction(
+                        getDistance(activeTripSegment.end, new Coordinates(nextStep)),
+                        nextStep.relativeDirection
```
If this is a turn instruction, it should include the name of the path/street.

binh-dam-ibigroup commented on src/test/java/org/opentripplanner/middleware/controllers/api/ManageLegTraversalTest.java:
```diff
@@ -118,20 +130,104 @@ private static Stream<Arguments> createTrace() {
+            bearing = bearing - 180;
+        }
+        Coordinates end = createDestinationPoint(point,distance, bearing);
+        Coordinates start = createDestinationPoint(point,distance + 0.005, bearing);
```
Insert space before `distance`.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/triptracker/TripInstruction.java:
```diff
@@ -1,30 +1,217 @@
+                return IMMEDIATE_PREFIX + stepInstruction;
+            } else if (distance <= UPCOMING_INSTRUCTION_RADIUS_DISTANCE) {
+                return UPCOMING_PREFIX + stepInstruction;
+            }
```
What do you think of setting the instruction to `DEVIATED` (instead of `NO_INSTRUCTION`) if the distance is too far?

br648 replied:
I associate deviated with trip tracking and not trip instruction. If you feel strongly about this, I'll change it. It would also be useful to get feedback from live testing.

binh-dam-ibigroup replied:
I think we can defer that for now.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/triptracker/TripInstruction.java:
```diff
@@ -1,30 +1,217 @@
+     */
+    private static String approachingStep(TripSegment activeTripSegment, StepSegment nearestStep, Leg expectedLeg) {
+        return getInstruction(
+            getDistance(activeTripSegment.end, nearestStep.start),
```
I am not sure about what `activeTripSegment` represents when calling the `/starttracking` endpoint. With a walk-to-transit itinerary, when calling the `/starttracking` endpoint with the start location of the first walk step, the distance computed seems to use the ending point of the `activeTripSegment` which can be away from the starting location, whereas intuitively, the distance should be with respect to the starting location of the first step. Could you clarify?

br648 replied:
If I have understood correctly, the preceeding method `isTripSegmentNearerToDestination` calculates distance to destination based on start coordinates, the distance calculated here should also use the (activeTripSegment) start coordinates.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/triptracker/StepSegment.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    public StepSegment(Step startStep, int stepIndex) {
+        super(new Coordinates(startStep), null);
```
Any reason why the ending coordinate is left null?

binh-dam-ibigroup reviewed (commented):
More comments. Could you clarify what should be happening for the first step of a walt-to-transit itinerary and the last step of a transit-to-walk itinerary?

binh-dam-ibigroup reviewed (changes requested):
The main ask for now, per https://github.com/ibi-group/otp-middleware/pull/223#discussion_r1539451705, is to add the street name to the turn instruction.

br648 committed:
refactor(Addessed PR feedback):

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/triptracker/StepSegment.java:
```diff
@@ -13,14 +13,28 @@ public class StepSegment extends Segment {
-    public StepSegment(Step startStep, int stepIndex) {
-        super(new Coordinates(startStep), null);
+    /**
+     * Used to hold the nearest step to the traveler's position, and it's index. The end coordinate and distance are
```
Typo "**its** index"

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/triptracker/TripInstruction.java:
```diff
@@ -110,36 +114,41 @@ private static boolean isLastStep(int stepIndex, Leg expectedLeg) {
+     * <p>
+     * "CONTINUE on Langley Drive"
+     * "RIGHT on service road"
+     * "Head WEST on sidewalk approx 6 meters"
```
Add TODO comment for internationalization and refinements to these generated instructions. Alternately, I can walk you through that and perhaps do that in this PR if the changes aren't too many.

binh-dam-ibigroup replied:
Actually, just add a TODO comment. We'll need to coordinate with the mobile app folks to see what they will need in the end.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/utils/Coordinates.java:
```diff
@@ -58,6 +58,9 @@ public int hashCode() {
     @Override
     public String toString() {
-        return lat + "," + lon + ",";
+        return "Coordinates{" +
```
Use `String.format`

br648 replied:
I've made this update, but this is an interesting case. The toString was generated by Intellij!

binh-dam-ibigroup commented on src/test/java/org/opentripplanner/middleware/controllers/api/ManageLegTraversalTest.java:
```diff
@@ -157,77 +156,73 @@ private static Stream<Arguments> createTurnByTurnTrace() {
-        Coordinates start = createDestinationPoint(point,distance + 0.005, bearing);
-        return new TripSegment(start, end);
+        Coordinates end = createDestinationPoint(point, distanceInMeters, bearing);
+        Coordinates start = createDestinationPoint(point,distanceInMeters + 5, bearing);
```
Insert space before `distanceInMeters`

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/triptracker/TripInstruction.java:
```diff
@@ -1,30 +1,226 @@
+            String relativeDirection = (step.relativeDirection.equals("DEPART"))
+                ? "Head " + step.absoluteDirection
+                : step.relativeDirection;
+            return (Math.round(distance) == 0)
```
We need a threshold here. There is no way someone's phone will report the exact spot from the itinerary.

br648 replied:
I've reinstated `IMMEDIATE_INSTRUCTION_RADIUS_DISTANCE` with a value of 2 meters.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/triptracker/TripInstruction.java:
```diff
@@ -110,36 +114,41 @@ private static boolean isLastStep(int stepIndex, Leg expectedLeg) {
-                return IMMEDIATE_PREFIX + stepInstruction;
-            } else if (distance <= UPCOMING_INSTRUCTION_RADIUS_DISTANCE) {
-                return UPCOMING_PREFIX + stepInstruction;
-            }
```
This was previously working. I am not able to get an instruction except for the first move after this last commit.

br648 replied:
Hopefully the latest updates should address this.

br648 replied:
@binh-dam-ibigroup is this now working again?

binh-dam-ibigroup reviewed (commented):
Can't get the instructions to appear when nearing a turn, except for the first instruction... Can you advise?

br648 committed:
refactor(Addressed PR feedback and introduced Aligned Step):

br648 committed:
refactor(Various changes to traveler locating):

br648 committed:
refactor(Removed step segment):

binh-dam-ibigroup commented on src/main/resources/env.schema.json:
```diff
@@ -257,6 +257,31 @@
+    "TRIP_INSTRUCTION_IMMEDIATE_DISTANCE": {
+      "type": "integer",
+      "examples": ["2"],
+      "description": "The distance under which an immediate instruction is given."
```
In which unit is the distance (2 instances)?

binh-dam-ibigroup commented on src/test/java/org/opentripplanner/middleware/controllers/api/TrackedTripControllerTest.java:
```diff
@@ -18,7 +18,7 @@
 import org.opentripplanner.middleware.triptracker.ManageTripTracking;
 import org.opentripplanner.middleware.triptracker.TrackingLocation;
-import org.opentripplanner.middleware.triptracker.TripInstruction;
+import org.opentripplanner.middleware.triptracker.TravelerLocator;
```
This import seems unused.

binh-dam-ibigroup commented on src/test/java/org/opentripplanner/middleware/controllers/api/TrackedTripControllerTest.java:
```diff
@@ -42,6 +42,7 @@
 import static org.opentripplanner.middleware.testutils.ApiTestUtils.TEMP_AUTH0_USER_PASSWORD;
 import static org.opentripplanner.middleware.testutils.ApiTestUtils.getMockHeaders;
 import static org.opentripplanner.middleware.testutils.ApiTestUtils.makeRequest;
+import static org.opentripplanner.middleware.triptracker.TripInstruction.NO_INSTRUCTION;
```
This import seems unused.

binh-dam-ibigroup commented:
~Not sure what you changed. When I call `starttracking` with coordinates very close (to the 6th decimal point) to the itinerary starting location, I get `DEVIATED` which is incorrect, and `NO_INSTRUCTION`, which will not be useful to users:~
```
{
  "frequencySeconds": 5,
  "instruction": "NO_INSTRUCTION",
  "journeyId": "xxxxx",
  "tripStatus": "DEVIATED",
  "message": null
}
```

binh-dam-ibigroup commented on README.md:
```diff
@@ -269,6 +269,11 @@ The special E2E client settings should be defined in `env.yml`:
 | TRIP_TRACKING_SUBWAY_BOUNDARY | integer | Optional | 100 | The threshold in meters below which travelling by subway is considered on track. |
 | TRIP_TRACKING_TRAM_BOUNDARY | integer | Optional | 100 | The threshold in meters below which travelling by tram is considered on track. |
 | TRIP_TRACKING_RAIL_BOUNDARY | integer | Optional | 100 | The threshold in meters below which travelling by rail is considered on track. |
+| TRIP_INSTRUCTION_ARRIVED_PREFIX | string | Optional | ARRIVED:  | The prefix to use when at destination. |
+| TRIP_INSTRUCTION_IMMEDIATE_DISTANCE | integer | Optional | 2 | The distance under which an immediate instruction is given. |
+| TRIP_INSTRUCTION_IMMEDIATE_PREFIX | string | Optional | IMMEDIATE:  | The prefix to use when on an instruction. |
+| TRIP_INSTRUCTION_UPCOMING_DISTANCE | integer | Optional | 10 | The distance under which an upcoming instruction is given. |
+| TRIP_INSTRUCTION_UPCOMING_PREFIX | string | Optional | UPCOMING:  | The prefix to use when nearing an instruction. |
```
Hmm, except for distance thresholds, don't make these parameters configurable. I sense this stuff will have to eventually be in the language bundle instead.

br648 replied:
Removed.

binh-dam-ibigroup commented on README.md:
```diff
@@ -269,6 +269,11 @@ The special E2E client settings should be defined in `env.yml`:
 | TRIP_TRACKING_TRAM_BOUNDARY | integer | Optional | 100 | The threshold in meters below which travelling by tram is considered on track. |
 | TRIP_TRACKING_RAIL_BOUNDARY | integer | Optional | 100 | The threshold in meters below which travelling by rail is considered on track. |
+| TRIP_INSTRUCTION_ARRIVED_PREFIX | string | Optional | ARRIVED:  | The prefix to use when at destination. |
+| TRIP_INSTRUCTION_IMMEDIATE_DISTANCE | integer | Optional | 2 | The distance under which an immediate instruction is given. |
```
Specify the unit for the distance thresholds. Are distances a radius or diameter from a turn point?

binh-dam-ibigroup commented on configurations/default/env.yml.tmp:
```diff
@@ -76,12 +76,23 @@ VALIDATE_ENVIRONMENT_CONFIG: true
 # The expected frequency to receive live journey location data.
 TRIP_TRACKING_UPDATE_FREQUENCY_SECONDS: 5
-# The minimum segment size in seconds for interpolated points.
+# The minimum trip segment size in seconds for interpolated points.
```
size => duration

binh-dam-ibigroup commented on configurations/default/env.yml.tmp:
```diff
@@ -76,12 +76,23 @@ VALIDATE_ENVIRONMENT_CONFIG: true
+
+# The prefix to use when at destination.
+TRIP_INSTRUCTION_ARRIVED_PREFIX
+# The distance under which an immediate instruction is given.
```
See my comment in README about these parameters.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/triptracker/LegSegment.java:
```diff
@@ -0,0 +1,40 @@
+
+    @Override
+    public String toString() {
+        return "LegSegment{" +
```
Maybe use String.format

br648 replied:
Done. Interesting: https://stackoverflow.com/questions/35933114/intellij-idea-template-for-tostring-method-using-string-format-instead-of-concat

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java:
```diff
@@ -0,0 +1,160 @@
+            travelerPosition.expectedLeg != null &&
+            travelerPosition.expectedLeg.mode.equalsIgnoreCase("walk")
+        ) {
+            TripInstruction tripInstruction = alignTravelerToTrip(travelerPosition, isStartOfTrip);
```
There is one case where a turn instruction is skipped if a leg segment is very short: (see that `prefix` is null but there is a turn instruction in the leg step.
<img width="624" alt="image" src="https://github.com/ibi-group/otp-middleware/assets/56846598/b300c98d-a3f1-4ea8-9a39-36049096de26">

binh-dam-ibigroup replied:
This is the itinerary where there is a left turn immediately folllowed by the right turn picture above: https://atlrides.com/#/?ui_activeSearch=lljkknis4&ui_activeItinerary=0&fromPlace=805%20Adair%20Avenue%20NE%2C%20Atlanta%2C%20GA%2C%20USA%3A%3A33.77990266501841%2C-84.36094356729174&toPlace=Monroe%20Dr%20NE%20at%20Yorkshire%20Rd%20NE%3A%3A33.79570180624357%2C-84.36787918489891&date=2024-04-05&time=14%3A34&arriveBy=false&mode=WALK%2CTRANSIT

br648 replied:
This might be down to configuration. The distance to the turn from the traveler is 11.3 meters which is outside of the `UPCOMING` radius.

br648 replied:
To be tested again by Binh.

binh-dam-ibigroup replied:
This is still happening, I cannot find a spot that triggers the turn instruction.

binh-dam-ibigroup replied:
Okay, I got the turn instruction for this one, that was a tricky one to find!

binh-dam-ibigroup replied:
I think regardless that this will be an issue when there is winding routing such as in the screenshot below. Just using the distance on line 129 in this file will cause the logic to skip turn steps such as the one circled in the screenshot, and that will be an issue later.

<img width="320" alt="image" src="https://github.com/ibi-group/otp-middleware/assets/56846598/21df4ee9-3795-4e05-91a4-b2257f45680c">

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java:
```diff
@@ -0,0 +1,160 @@
+    /**
+     * Determine if the step is near to the destination than the traveler.
+     */
+    private static boolean isStepNearerToDestination(
```
This might be tricky. When you have to go around an impassable place to reach your destination, this method will think you are walking in the wrong direction. Would it be easier to just reference the step order within the leg?

br648 replied:
@binh-dam-ibigroup The method name is a bit misleading so I have updated it to `isFirstPositionNearestToTarget`. I have also moved the method into the `GeometryUtils` class. The trip destination is only used if working if the last step, in all other cases the next step in the leg is used i.e. the step order within the leg.

binh-dam-ibigroup commented:
> When I call `starttracking` with coordinates very close (to the 6th decimal point) to the itinerary starting location, I get `DEVIATED` which is incorrect...

Disregard. It can be tricky to find the starting location.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/triptracker/TravelerPosition.java:
```diff
@@ -0,0 +1,45 @@
+        TrackingLocation lastLocation = trackedJourney.locations.get(trackedJourney.locations.size() - 1);
+        currentTime = lastLocation.timestamp.toInstant();
+        currentPosition = new Coordinates(lastLocation);
+        expectedLeg = getExpectedLeg(lastLocation.timestamp.toInstant(), itinerary);
+        legSegmentFromTime = getSegmentFromTime(lastLocation.timestamp.toInstant(), itinerary);
```
Reuse `currentTime`.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/triptracker/TripInstruction.java:
```diff
@@ -1,30 +1,92 @@
+
+public class TripInstruction {
+
+    /** The distance under which an immediate instruction is given. */
```
See README comment regarding the unit for the distance parameters.

binh-dam-ibigroup commented on src/test/java/org/opentripplanner/middleware/controllers/api/ManageLegTraversalTest.java:
```diff
@@ -118,20 +126,125 @@ private static Stream<Arguments> createTrace() {
+    private static Stream<Arguments> createTurnByTurnTrace() {
+        List<Step> walkSteps = busStopToJusticeCenterItinerary.legs.get(0).steps;
+        Coordinates originCoords = new Coordinates(busStopToJusticeCenterItinerary.legs.get(0).from);
+        Step stepOne = walkSteps.get(0);
+        Coordinates stepOneCoords = new Coordinates(stepOne);
+        Step stepTwo = walkSteps.get(1);
+        Coordinates stepTwoCoords = new Coordinates(stepTwo);
+        Step stepThree = walkSteps.get(2);
+        Coordinates stepThreeCoords = new Coordinates(stepThree);
+        Step stepFour = walkSteps.get(3);
+        Coordinates stepFourCoords = new Coordinates(stepFour);
+        Step stepFive = walkSteps.get(4);
+        Coordinates stepFiveCoords = new Coordinates(stepFive);
```
Can these things be an array or a list instead? It will be easier to build your arguments stream from that.

br648 replied:
I thought this was easier to follow by naming the steps and coords. The arguments don't always use the step with it's coords, e.g. "Approaching the destination." which is a combination of the last step and destination coords. Lemme know what you think, I may have misunderstood.

binh-dam-ibigroup commented on src/test/java/org/opentripplanner/middleware/controllers/api/ManageLegTraversalTest.java:
```diff
@@ -118,20 +126,125 @@ private static Stream<Arguments> createTrace() {
-            Segment segmentTwo = segments.get(i+1);
-            assertEquals(segmentOne.end.lat, segmentTwo.start.lat);
+        List<LegSegment> legSegments = createSegmentsForLeg();
+        for (int i = 0; i < legSegments.size()-1; i++) {
+            LegSegment legSegmentOne = legSegments.get(i);
+            LegSegment legSegmentTwo = legSegments.get(i+1);
```
Insert spaces between pluses and minuses.

binh-dam-ibigroup commented on src/test/java/org/opentripplanner/middleware/controllers/api/ManageLegTraversalTest.java:
```diff
@@ -140,74 +253,99 @@ void canTrackLegWithoutDeviating() {
+    @Test
+    void canTrackTripTurnByTurn() {
+        ZonedDateTime startOfTrip = ZonedDateTime.ofInstant(
+            busStopToJusticeCenterItinerary.legs.get(0).startTime.toInstant(),
```
Extract var for `busStopToJusticeCenterItinerary.legs.get(0)` and its first step.

br648 replied:
Sorry, I was meaning to remove this unless unit test.

binh-dam-ibigroup reviewed (changes requested):
The logic is much better this time, nice work! I think we are now in the minor tweaking phase. I'm going to ask you take the `PREFIX` params out of the configuration and make them constants for now, with the intent to move them to the i18n bundle in subsequent work, and specify in the docs (or in the parameter names) the units in which distances are expressed. Fixing the missed turn case is probably a nice-to-have (unless the second reviewer finds that blocking). See other refactoring suggestions.

br648 committed:
refactor(Addressed PR feedback):

br648 committed:
refactor(Refactored to include instruction for deviated trip status):

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/triptracker/ManageTripTracking.java:
```diff
@@ -48,13 +42,14 @@ public static StartTrackingResponse startTracking(Request request) {
 
                 try {
-                    TripStatus status = getTripStatus(trackedJourney, monitoredTrip.itinerary);
+                    TravelerPosition travelerPosition = new TravelerPosition(trackedJourney, monitoredTrip.itinerary);
```
This needs to be `monitoredTrip.journeyState.matchingItinerary`.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/triptracker/ManageTripTracking.java:
```diff
@@ -83,12 +78,12 @@ public static UpdateTrackingResponse updateTracking(Request request) {
 
                     try {
-                        TripStatus status = getTripStatus(trackedJourney, monitoredTrip.itinerary);
+                        TravelerPosition travelerPosition = new TravelerPosition(trackedJourney, monitoredTrip.itinerary);
```
This needs to be `monitoredTrip.journeyState.matchingItinerary`.

binh-dam-ibigroup replied:
The reason for that is that the itinerary matches for the recurring day of the monitored trip.

binh-dam-ibigroup commented:
@br648 Your last changes got instruction to head to the desired street if deviated, however I can't get upcoming turn directions when updating a location to near a turn.

binh-dam-ibigroup reviewed (changes requested):
I can't get turn instructions to show anymore. Any reason for that?

br648 committed:
refactor(Fixed issue with trip instruction): Correctly set trip instruction to on track

br648 commented:
> I can't get turn instructions to show anymore. Any reason for that?

@binh-dam-ibigroup Sorry! I hadn't correctly set the "tripOnTrack" value. See: https://github.com/ibi-group/otp-middleware/pull/223/commits/34fd4fcd984f316c3eddc2dd5d217f220b567767

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java:
```diff
@@ -0,0 +1,178 @@
+                : null;
+        }
+
+        Coordinates destination = new Coordinates(travelerPosition.expectedLeg.to);
```
Rename to `legDestination`?

binh-dam-ibigroup reviewed (commented):
More comments regarding J-shaped itineraries.

br648 committed:
refactor(Injected steps into leg geo points): Next step is now found by traversing a list of geo poi

A travelers position is snapped to a leg geo point. The next step is found from this point index
forward and not by distance.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java:
```diff
@@ -73,12 +75,10 @@ private static boolean hasRequiredTripStatus(TripStatus tripStatus) {
-            return null;
-        }
-        return new TripInstruction(nearestStep.streetName);
+        Step nearestStep = snapToLeg(travelerPosition);
```
Maybe rename to `snapToStep`.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java:
```diff
@@ -99,80 +99,113 @@ public static TripInstruction alignTravelerToTrip(TravelerPosition travelerPosit
+            for (Step step : leg.steps) {
+                if (isPointBetween(p1, p2, new Coordinates(step)) && !injectedSteps.contains(step)) {
+                    finalPositions.add(new Coordinates(step));
+                    injectedSteps.add(step);
```
That seems clever

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java:
```diff
@@ -99,80 +99,113 @@ public static TripInstruction alignTravelerToTrip(TravelerPosition travelerPosit
+            }
+
+        }
+        return finalPositions;
```
Some consecutive coordinates are duplicated, but it's probably fine to keep them.

<img width="516" alt="image" src="https://github.com/ibi-group/otp-middleware/assets/56846598/a66a84a8-7ad9-401e-941e-0fbf527babc8">

br648 replied:
It shouldn't impact functionality, but I have removed them and made this part a bit cleaner.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/triptracker/TravelerLocator.java:
```diff
@@ -99,80 +99,113 @@ public static TripInstruction alignTravelerToTrip(TravelerPosition travelerPosit
+                    finalPositions.add(pointIndex, new Coordinates(missedStep));
+                }
+            }
+
```
Remove blank line.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/triptracker/TripStatus.java:
```diff
@@ -65,23 +64,24 @@ public enum TripStatus {
             }
-            if (segmentFromPosition != null && isWithinModeBoundary(currentPosition, segmentFromPosition)) {
-                Instant segmentCenterTime = expectedLeg
+            if (travelerPosition.legSegmentFromPosition != null &&
```
I would be in favor of putting the position first, then time after. If someone is ahead or behind schedule on a multi-leg itinerary, we wouldn't be able to get the correct leg or step they are at using this code.

br648 replied:
@binh-dam-ibigroup This makes sense and I've made the change. It does mean that `legSegmentFromTime` will be redundant because a lat/lon will always be attributed to a leg segment regardless of how far it is from it. Any thoughts on this before I remove the related code?

binh-dam-ibigroup replied:
I think it will be fine to remove.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/triptracker/TripInstruction.java:
```diff
@@ -1,30 +1,118 @@
+
+    /** The radius in meters under which an immediate instruction is given. */
+    public static final int TRIP_INSTRUCTION_IMMEDIATE_RADIUS
+        = getConfigPropertyAsInt("TRIP_INSTRUCTION_IMMEDIATE_RADIUS", 2);
```
I feel this will have to be increased, but we will tweak that separately.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/triptracker/TripInstruction.java:
```diff
@@ -1,30 +1,118 @@
+    public static final int TRIP_INSTRUCTION_UPCOMING_RADIUS
+        = getConfigPropertyAsInt("TRIP_INSTRUCTION_UPCOMING_RADIUS", 10);
+
+    /** The prefix to use when on an instruction. */
```
```suggestion
    /** The prefix to use when at a street location with an instruction. */
```

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/triptracker/TripInstruction.java:
```diff
@@ -1,30 +1,118 @@
+    /** The prefix to use when on an instruction. */
+    public static final String TRIP_INSTRUCTION_IMMEDIATE_PREFIX = "IMMEDIATE: ";
+
+    /** The prefix to use when nearing an instruction. */
```
```suggestion
    /** The prefix to use when nearing a street location with an instruction. */
```

binh-dam-ibigroup reviewed (changes requested):
Nice job handling the J-shaped routing. One small change regarding doing prioritizing obtaining the leg from the location over obtaining the leg using time.

br648 committed:
Update src/main/java/org/opentripplanner/middleware/triptracker/TripInstruction.java

Co-authored-by: Binh Dam <56846598+binh-dam-ibigroup@users.noreply.github.com>

br648 committed:
Update src/main/java/org/opentripplanner/middleware/triptracker/TripInstruction.java

Co-authored-by: Binh Dam <56846598+binh-dam-ibigroup@users.noreply.github.com>

br648 committed:
refactor(Addressed PR feedback): Trip status is now determined by position and then time

br648 committed:
refactor(Removed func related to getting leg segment from time):

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/triptracker/TripStatus.java:
```diff
@@ -66,33 +66,34 @@ public enum TripStatus {
-                return travelerPosition.currentTime.isBefore(segmentCenterTime)
-                    ? TripStatus.AHEAD_OF_SCHEDULE
-                    : TripStatus.BEHIND_SCHEDULE;
+                    .plusSeconds((long) getSegmentStartTime(travelerPosition.legSegmentFromPosition));
```
What was the reason to change from `getSegmentTimeInterval` to `getSegmentStartTime`? This change seems to report me "ahead of schedule" even if I am behind schedule.

binh-dam-ibigroup replied:
Oh I see, never mind. If I am way behind schedule, the trip monitoring might have switched to the next available trip the next day already...

br648 replied:
Just to add this. The method now has to cover on schedule so it needs to consider the start and end segment times.

br648 committed:
refactor(Created exclusion zone around steps):

binh-dam-ibigroup commented on configurations/test/env.yml:
```diff
@@ -51,4 +51,17 @@ MAXIMUM_PERMITTED_MONITORED_TRIPS: 5
+TRIP_TRACKING_TRAM_BOUNDARY: 100
+TRIP_TRACKING_RAIL_BOUNDARY: 200
+TRIP_INSTRUCTION_IMMEDIATE_PREFIX_DISTANCE: 2
+TRIP_INSTRUCTION_IMMEDIATE_PREFIX: "IMMEDIATE: "
```
Delete this setting (it is not configurable)

binh-dam-ibigroup commented on configurations/test/env.yml:
```diff
@@ -51,4 +51,17 @@ MAXIMUM_PERMITTED_MONITORED_TRIPS: 5
+TRIP_INSTRUCTION_IMMEDIATE_PREFIX_DISTANCE: 2
+TRIP_INSTRUCTION_IMMEDIATE_PREFIX: "IMMEDIATE: "
+TRIP_INSTRUCTION_UPCOMING_PREFIX_DISTANCE: 10
+TRIP_INSTRUCTION_UPCOMING_PREFIX: "UPCOMING: "
```
Delete this setting (it is not configurable)

binh-dam-ibigroup reviewed (approved):
Nice work, and the latest changes are so much better!
A couple of notes for subsequent work:
- To determine whether the user is on a walk leg that follows a transit/non-walk leg, the user's position should be used over the current time.
- If the user is within the "upcoming" radius of a turn, the turn instruction should probably be given even if the user is "Deviated" from the itinerary.

br648 committed:
refactor(env.yml): Removed redundant params

br648 committed:
refactor(Fixed merge conflicts):

br648 committed:
Merge branch 'dev' into feature/OTP-1050-turn-by-turn-processing

JymDyerIBI reviewed (approved):
This looks good, makes sense. I didn't replicate the `TravelerLocator` edge condition.
